firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/ApkHashExtractor.java

```diff
@@ -0,0 +1,140 @@
+package com.google.firebase.appdistribution.impl;
+
+import static com.google.firebase.appdistribution.impl.ReleaseIdentificationUtils.getPackageInfoWithMetadata;
+
+import android.content.Context;
+import android.content.pm.PackageInfo;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+import com.google.firebase.appdistribution.FirebaseAppDistributionException;
+import com.google.firebase.appdistribution.FirebaseAppDistributionException.Status;
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.Locale;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+/** Extracts a hash of the installed APK. */
+class ApkHashExtractor {
+
+  private static final String TAG = "ApkHashExtractor:";
+  private static final int BYTES_IN_LONG = 8;
+
+  private final ConcurrentMap<String, String> cachedApkHashes = new ConcurrentHashMap<>();
```

Comment:
Note that this cache used to be static (global). That didn't make sense given that the hash depends on the Context that's provided to each instance. Instead I've pulled out the creation of the ApkHashExtractor to the FirebaseAppDistributionRegistrar so that there is clearly only one of them.